src/types.rs

```diff
@@ -0,0 +1,348 @@
+            Type::FuncRef,
+        ];
+
+        VARIANTS.iter()
+    }
+}
+
+impl From<&Type> for &'static str {
+    fn from(value: &Type) -> Self {
+        match value {
+            Type::I32 => "I32",
+            Type::I64 => "I64",
+            Type::F32 => "F32",
+            Type::F64 => "F64",
+            Type::V128 => "V128",
+            Type::ExternRef => "EXTERN_REF",
+            Type::FuncRef => "FUNC_REF",
+        }
+    }
+}
+
+impl ToPyObject for Type {
+    fn to_object(&self, py: Python) -> PyObject {
+        (*self as u8).into_py(py)
+    }
+}
+
+impl IntoPy<PyObject> for Type {
+    fn into_py(self, py: Python) -> PyObject {
+        self.to_object(py)
+    }
+}
+
+impl<'source> FromPyObject<'source> for Type {
+    fn extract(obj: &'source PyAny) -> PyResult<Self> {
+        let variant = u8::extract(obj)?;
+
+        Ok(match variant {
+            1 => Self::I32,
+            2 => Self::I64,
+            3 => Self::F32,
+            4 => Self::F64,
+            5 => Self::V128,
+            6 => Self::ExternRef,
+            7 => Self::FuncRef,
+            _ => {
+                return Err(to_py_err::<ValueError, _>(
+                    "Failed to extract `Type` from `PyAny`",
+                ))
+            }
+        })
+    }
+}
+
+impl From<wasmer::Type> for Type {
+    fn from(value: wasmer::Type) -> Self {
+        match value {
+            wasmer::Type::I32 => Self::I32,
+            wasmer::Type::I64 => Self::I64,
+            wasmer::Type::F32 => Self::F32,
+            wasmer::Type::F64 => Self::F64,
+            wasmer::Type::V128 => Self::V128,
+            wasmer::Type::ExternRef => Self::ExternRef,
+            wasmer::Type::FuncRef => Self::FuncRef,
+        }
+    }
+}
+
+impl Into<wasmer::Type> for Type {
+    fn into(self) -> wasmer::Type {
+        match self {
+            Self::I32 => wasmer::Type::I32,
+            Self::I64 => wasmer::Type::I64,
+            Self::F32 => wasmer::Type::F32,
+            Self::F64 => wasmer::Type::F64,
+            Self::V128 => wasmer::Type::V128,
+            Self::ExternRef => wasmer::Type::ExternRef,
+            Self::FuncRef => wasmer::Type::FuncRef,
+        }
+    }
+}
+
+#[pyclass]
+pub struct FunctionType {
+    #[pyo3(get)]
+    pub params: Vec<Type>,
+
+    #[pyo3(get)]
+    pub results: Vec<Type>,
+}
+
+#[pymethods]
+impl FunctionType {
+    #[new]
+    fn new(params: Vec<Type>, results: Vec<Type>) -> Self {
+        Self { params, results }
+    }
+}
+
+impl From<&wasmer::FunctionType> for FunctionType {
+    fn from(value: &wasmer::FunctionType) -> Self {
+        Self {
+            params: value
+                .params()
+                .to_vec()
+                .into_iter()
+                .map(Into::into)
+                .collect(),
+            results: value
+                .results()
+                .to_vec()
+                .into_iter()
+                .map(Into::into)
+                .collect(),
```

Comment:
Minor efficiency thing, but I think `to_vec` before iterating probably isn't necessary

Reply:
I need to implement `From<&wasmer::Type> for Type` in this case. Not a big deal :-). Updating the code!

Reply:
You should also be able to used `cloned` I think!